### DISCUSSION TITLE
How to scale Canvas?

### DISCUSSION THREAD

**Discussion by Joshmamroud** [GrapesJS/grapesjs#3686] [Q&A]
What is the best way to scale the canvas similar to the way Google Docs does?
https://www.loom.com/share/a8935cde410042688e3d2f775de173fe

**Comment by artf** [comment]
You can use this [`Canvas.setZoom`](https://grapesjs.com/docs/api/canvas.html#setzoom)

**Reply by Joshmamroud** [reply]
@artf I did try that however it, as the name implies, just zooms out from the canvas. I would like to be able to zoom up or down and preserve the original height and size of the scroll bar.

Here's what it currently looks like zoomed to 50%

![image](https://user-images.githubusercontent.com/5224187/132127876-2b466c77-a7ce-4e21-b710-72f9fac71d3c.png)

**Reply by artf** [reply]
That happens because the frame, by default, sets the height of the available space, but you can change it if you need, eg.
```js
editor.Pages.getSelected().getMainFrame().set('height', '600px');
```

**Reply by Joshmamroud** [reply]
@artf I see. The issue with this method is it doesn't scale the contents of the frame, just reduces the available space. I'm starting to think it might not be possible? Or perhaps use a combination of the two and scale the scrollbar. Unless you have a better solution in mind. Any help is appreciated. Thanks!

**Reply by artf** [reply]
Well, I think the current implementation works until we don't introduce the live collaboration, where dealing with scrollable frames would be impossible for multiple users. So, one thing I've thought of is a new option for frames that would enable automatic height on content change and then make the whole canvas scrollable.
This approach actually works well for stuff like websites and mobile apps, but as you work with documents it might be not ideal, where you would need to place a height limit (height of the document) and generate a new page (frame) when the limit is exceeded (eg. similar to how Word works).
Anyway, did you try to change the frame height and then make the canvas element scrollable (via CSS). Maybe that would work for your case.

**Reply by Joshmamroud** [reply]
If we could scroll the canvas instead of the frame that would be huge! It's not really an issue with our use case since users are not actually creating the document itself but rather the template. This means that there's no way of knowing where the document page would end since it's dynamic and determined by the data being merged.

I haven't yet tested changing the frame height and making the canvas scrollable via CSS as I haven't had the opportunity. I will definitely test it out and report back, though.

**Reply by gxanshu** [reply]
> Well, I think the current implementation works until we don't introduce the live collaboration, where dealing with scrollable frames would be impossible for multiple users. So, one thing I've thought of is a new option for frames that would enable automatic height on content change and then make the whole canvas scrollable. This approach actually works well for stuff like websites and mobile apps, but as you work with documents it might be not ideal, where you would need to place a height limit (height of the document) and generate a new page (frame) when the limit is exceeded (eg. similar to how Word works). Anyway, did you try to change the frame height and then make the canvas element scrollable (via CSS). Maybe that would work for your case.

i also working on a project where i use grapesjs to design a PDF document i need to place a height limit (height of the document) and generate a new page (frame) when the limit is exceeded. how can i achieve something like that ?

**Comment by badiuciprian** [comment]
I don't know if this is the best approach, but I didi it by forcing css styles to the gjs-frame-wrapper and tools. I used a slider to set the value.  

 const onChange = (newValue: number) => {
    setZoom(newValue as any);
    const frame = document.querySelector<HTMLElement>(`#gjs-${projectFile._id} .gjs-frame-wrapper`);
    const tools = document.querySelector<HTMLElement>(`#gjs-${projectFile._id} .gjs-cv-canvas__tools`);
    const handles = document.querySelector<HTMLElement>(`#gjs-${projectFile._id} #gjs-tools`);
    frame.style.transform = 'scale(' + (newValue) + ')';
    frame.style.transformOrigin = 'top left';
    frame.style.transition = '0.2s';
    tools.style.transform = 'translate(15px, 15px) scale(' + (newValue) + ')';
    tools.style.transformOrigin = 'top left';
    handles.style.transform = 'translate(-15px, -15px)';
  };

I needed to correct the position of the handles because i had some other elements outside that caused the 15px offset from the parent
